In [3]:
import torch

In [7]:
x = 3.5
y = x*x + 2
print(x, y)
print(x*x)

3.5 14.25
12.25


In [9]:
a =torch.tensor(3.5)

In [11]:
print(a)

tensor(3.5000)


In [13]:
print(a+3)

tensor(6.5000)


In [15]:
x1 = torch.tensor(5.5)
y1 = x1+8
print(y1)

tensor(13.5000)


In [21]:
x = torch.tensor(3.5, requires_grad=True)
print("x: ", x)

x:  tensor(3.5000, requires_grad=True)


In [23]:
# b1 is defined as a function of a1
y = (x-1)*(x-2)*(x-3)
print(y)

tensor(1.8750, grad_fn=<MulBackward0>)


We can see that y is calculated from the expression (x-1) * (x-2) * (x-3).

As expected, the value of y is 
1.8750
1.8750
. This is because x is 
3.5
3.5
, so (3.5-1) * (3.5-2) * (3.5-3) is 
1.8750
1.8750
.

The following graph of y = (x-1) * (x-2) * (x-3) illustrates what we’ve just calculated.

#The error gradient is the rate at which the output error changes as a result of changing network link weights.
#Automatic differentiation and gradient calculation#
Let’s work out the gradient of y at x = 3.5. That is, 
d
y
d
x
dx
dy
​
 
 at x = 3.5.

To do this, PyTorch needs to look at y and see which tensors it depends on, as well as the mathematical form of that dependency. It can then do the calculus to work out 
d
y
d
x
dx
dy
​
 
.

The following code computes the gradient of y w.r.t x:

In [29]:
x = torch.tensor(3.5, requires_grad=True)

# y is defined as a function of x
y = (x-1) * (x-2) * (x-3)

# work out gradients
y.backward()

In [31]:
print("Numerical value of gradient:", x.grad)

Numerical value of gradient: tensor(5.7500)


 computational graph showing gradient dz/dx
 y = x2
 z = 2y+3
 calculus  chain rule => dz/dx = dz/dy*dy/dx
 dz/dx = ?
 z = 2(x2)+3
 dz/dx = 2.2x
 dz/dx = 4x
 x = 3.5
 dy/dx = 4*3.5 = 14

In [40]:
x = torch.tensor(3.5, requires_grad=True)
y = x*x
z= 2*y+3
print("x:", x)
print("y:", y)
print("z = 2*y+3 = ", z)

x: tensor(3.5000, requires_grad=True)
y: tensor(12.2500, grad_fn=<MulBackward0>)
z = 2*y+3 =  tensor(27.5000, grad_fn=<AddBackward0>)


In [42]:
# work out gradients
z.backward()
print("working out gradient dz/dx")
print("Gradient at x= 3.5", x.grad)

working out gradient dz/dx
Gradient at x= 3.5 tensor(14.)


The answer is 
14
14
, which is exactly what we worked out by hand above. This is great!

It is worth noting that the gradient value inside the 
x
x
 tensor relates to how 
z
z
 changes. That’s because we asked PyTorch to work backwards from z using z.backward(). So x.grad is 
d
z
d
x
dx
dz
​
 
 not 
d
y
d
x
dx
dy
​
 
.

In [45]:
a = torch.tensor(2.0, requires_grad=True)
b = torch.tensor(1.0, requires_grad=True)

x = 2*a+3*b
y = 5*a*a+3*b*b*b
z = 2*x+3*y

z.backward()
print("working out gradient dz/da")
print("Gradient at a=2.0" ,a.grad)
print("working out gradient dz/db")
print("Gradient at b=1.0" ,b.grad)

working out gradient dz/da
Gradient at a=2.0 tensor(64.)
working out gradient dz/db
Gradient at b=1.0 tensor(33.)
